### ISSUE TITLE
Cannot reproduce the result (such as WER=0.133...)

### ISSUE THREAD

**Issue by SakuyaTachibana** [Fengdalu/LipNet-PyTorch#8]
Dear Fengdalu,

Thank you for sharing your LipNet codes as open source, which is very helpful for me.
I tried to reproduce your result (such as WER=0.133...) from scratch.
Concretely, I cloned your github repository and run "main.py" on our server, but I cannot obtain such a low WER scores.
(My WER scores is around 0.16 ...)

I have checked the old issue "How many epoch did you train? #1", and tuned the hyper-parameters (i.e., "option.py") based on your comment.
However, I cannot still obtain your result (such as WER=0.133...).
Do you have some idea why I cannot reproduce your result?
I'd appreciate it if you could give me some advice about that.

Best regards,
SakuyaTachibana

**Comment by Fengdalu** [comment]
Reduce learning rate may help. I reduce the learning rate from 1e-4 to 5e-5, 3e-5  when the model reach the bottleneck

**Comment by SakuyaTachibana** [comment]
Dear Fengdalu,

Thank you for your quick reply.
I am stuck in this problem for a long time...
I would be glad if you give me some additional advice about this problem.

How do you control the learning rate, e.g., manually or torch.optim.lr_scheduler?
I'd appreciate it if you share some codes (or advice) to control the learning rate?

Best regards,
SakuyaTachibana

**Comment by Fengdalu** [comment]
I recommend manually. In this project, I use tensorboard to visualize the train_loss / val_loss / val wer. When the train loss does not reduce in several epochs, it might be a suitable time to reduce LR.

**Comment by SakuyaTachibana** [comment]
Firstly, I am sorry for my long comment...

Thank you very much for your advice, again.
I will check the tensorboard.

----------
I am sorry for my many questions, but I would like to confirm two more questions.

(1) When you restart the training from the previous model by changing the learning rate (Is this manual control, right?), how do you deal with the Adam's statistics (e.g., gradient history), "keep the statistics from the previous optimizer" or "initialize the statistics from scratch"?

(2) If you remember about the previous experiment, it would be very helpful for me to share your experience, about what epoch you reduce the learning rate.
In the previous issue "How many epoch did you train? #1", you described that it requires nearly 200 - 300 epochs with lr=1e-4.
Does this means that we require about 200 - 300 epochs also for lr=5e-5 and lr=3e-5?
Totally, are about 600 - 900 epochs required?

**Comment by Fengdalu** [comment]
> Firstly, I am sorry for my long comment...
> 
> Thank you very much for your advice, again.
> I will check the tensorboard.
> 
> I am sorry for my many questions, but I would like to confirm two more questions.
> 
> (1) When you restart the training from the previous model by changing the learning rate (Is this manual control, right?), how do you deal with the Adam's statistics (e.g., gradient history), "keep the statistics from the previous optimizer" or "initialize the statistics from scratch"?
> 
> (2) If you remember about the previous experiment, it would be very helpful for me to share your experience, about what epoch you reduce the learning rate.
> In the previous issue "How many epoch did you train? #1", you described that it requires nearly 200 - 300 epochs with lr=1e-4. Does this means that we require about 200 - 300 epochs also for lr=5e-5 and lr=3e-5? Totally, are about 600 - 900 epochs required?

(1) In this project, all statistics from scratch if training is resumed after changing. If you want to save the state of the optimizer, you can dump the optimizer as a pkl using `torch.save`.

(2) I use 1e-4 for ~200 epochs and 5e-5, 3e-5... for several iterations, the WER will not go better if the LR is too small.

**Comment by SakuyaTachibana** [comment]
Thank you for your response.
Based on your advice, I will try to tune my model.

Sorry, I have one more question.
I'd appreciate it if you give me an additional advice.

Following your advice, I checked tensorboard
Then, I found that the train_loss continues to reduce over 1000 epochs, while the valid_loss continues to increase since ~200 epoch.
It seems to be overfitting problem...
Is this common behavior with your experience?

**Comment by Fengdalu** [comment]
> Thank you for your response.
> Based on your advice, I will try to tune my model.
> 
> Sorry, I have one more question.
> I'd appreciate it if you give me an additional advice.
> 
> Following your advice, I checked tensorboard
> Then, I found that the train_loss continues to decrease over 1000 epochs, while the valid_loss continues to increase since ~200 epoch.
> It seems to be overfitting problem...
> 
> Is this common behavior with your experience?
> Is it possible to share your tensorboard for successful case?

My original setting is missed. In my experiment, the model will be saved and test every 100 iters, so it is easy to find the best point before overfitting.
